harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +34,107 @@ class DownloadMode(enum.Enum):
     NoSharedDownload = "NO_SHARED_DOWNLOAD"
 
 
+def merge_metadata(all_metadata: List[Dict[str, Any]]) -> Tuple[Dict[str, Any], Dict[str, List]]:
+    """
+      Given a list of metadata, return:
+        - merged metadata
+        - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
```

Comment:
I really appreciate detailed comments for complex behaviors :tada:. This code would be wicked hard to review without this comment.